core/src/main/java/org/elasticsearch/script/ScriptMetaData.java

```diff
@@ -0,0 +1,292 @@
+                    key = parser.currentName();
+                    break;
+                case START_OBJECT:
+                    XContentBuilder contentBuilder = XContentBuilder.builder(parser.contentType().xContent());
+                    contentBuilder.copyCurrentStructure(parser);
+                    scripts.put(key, new ScriptAsBytes(contentBuilder.bytes()));
+                    break;
+                default:
+                    throw new ParsingException(parser.getTokenLocation(), "Unexpected token [" + token + "]");
+            }
+        }
+        return new ScriptMetaData(scripts);
+    }
+
+    @Override
+    public EnumSet<MetaData.XContentContext> context() {
+        return MetaData.API_AND_GATEWAY;
+    }
+
+    @Override
+    public ScriptMetaData readFrom(StreamInput in) throws IOException {
+        int size = in.readVInt();
+        Map<String, ScriptAsBytes> scripts = new HashMap<>();
+        for (int i = 0; i < size; i++) {
+            String languageAndId = in.readString();
+            BytesReference script = in.readBytesReference();
+            scripts.put(languageAndId, new ScriptAsBytes(script));
+        }
+        return new ScriptMetaData(scripts);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        for (Map.Entry<String, ScriptAsBytes> entry : scripts.entrySet()) {
+            builder.rawField(entry.getKey(), entry.getValue().script);
+        }
+        return builder;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeVInt(scripts.size());
+        for (Map.Entry<String, ScriptAsBytes> entry : scripts.entrySet()) {
+            out.writeString(entry.getKey());
+            entry.getValue().writeTo(out);
+        }
+    }
+
+    @Override
+    public Diff<MetaData.Custom> diff(MetaData.Custom before) {
+        return new ScriptMetadataDiff((ScriptMetaData) before, this);
+    }
+
+    @Override
+    public Diff<MetaData.Custom> readDiffFrom(StreamInput in) throws IOException {
+        return new ScriptMetadataDiff(in);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        ScriptMetaData other = (ScriptMetaData) o;
+        return scripts.equals(other.scripts);
+    }
+
+    @Override
+    public int hashCode() {
+        return scripts.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return "ScriptMetaData{" +
+                "scripts=" + scripts +
+                '}';
+    }
+
+    static String toKey(String language, String id) {
+        if (id.contains("#")) {
+            throw new IllegalArgumentException("stored script id can't contain: '#'");
+        }
+        if (language.contains("#")) {
+            throw new IllegalArgumentException("stored script language can't contain: '#'");
+        }
```

Comment:
I'll add this validation also to `PutStoredScriptRequest`, but I like to keep it here too.
This is where the actual logic is and the validation should be next to it.